ipa/ipa_provider.py

```diff
@@ -447,6 +449,71 @@ def install_package(self, client, package):
                 log_file.write('\n')
                 log_file.write(out)
 
+    def process_injection_file(self, client):
+        """
+        Load yaml file and process injection configuration.
+
+        There are 5 injection options:
+
+        :inject_packages: an rpm path or list of rpm paths which will be
+                          copied and installed on instance.
+        :inject_archives: an archive or list of archives which will
+                          be copied and extracted on instance.
+        :inject_files: a file path or list of file paths which
+                       will be copied to instance.
+        :execute: a command or list of commands to run.
+        :install: a package name or list of package names to
+                 install from an existing repo.
+        """
+        configuration = ipa_utils.get_yaml_config(self.inject)
+
+        if configuration.get('inject_packages'):
+            inject_packages = configuration['inject_packages']
+
+            if not isinstance(inject_packages, list):
+                inject_packages = [inject_packages]
+
+            for package in inject_packages:
+                package_path = self.put_file(client, package)
+                self.install_package(client, package_path)
+
+        if configuration.get('inject_archives'):
+            inject_archives = configuration['inject_archives']
+
+            if not isinstance(inject_archives, list):
+                inject_archives = [inject_archives]
+
+            for archive in inject_archives:
+                archive_path = self.put_file(client, archive)
+                self.extract_archive(client, archive_path)
+
+        if configuration.get('inject_files'):
+            inject_files = configuration['inject_files']
+
+            if not isinstance(inject_files, list):
+                inject_files = [inject_files]
+
+            for file_path in inject_files:
+                self.put_file(client, file_path)
```

Comment:
One could, and some people will argue that packages and tar archives are files as well. I am OK leaving this option as if one want to inject a config file that effects a package forcing them to create an archive is silly and it saves us the logic of having to determine file types for certain operation.

Reply:
Yeah "put" is the first step to injecting a package and an archive. So exposing it independently I think is nice. I.e. like your example to copy a single file.